# **VGG Preprocessor - Autonomous Vehicle Grand Theft Auto V**

This notebook contains the input data preprocessor for VGG based model for autonomous vehicle project. It saves all processed data in Numpy arrays, so it can be used by the model.

Imports general purpose libraries

In [ ]:
import numpy as np
import h5py
import matplotlib.pyplot as plt
import keras
import shutil
import os
import random

Using TensorFlow backend.


## **Getting data**

Mounts Google Drive in Google Colab disk to get data

In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


Loads all training data from Google Drive folder

In [ ]:
Y=np.load('/content/drive/My Drive/Novos_Dados/YTreino0.npy')
print("Arquivo 0 carregado")
for i in range(1,35):
    Y=np.concatenate((Y, np.load('/content/drive/My Drive/Novos_Dados/YTreino'+str(i)+'.npy')))
    print("Arquivo %d carregado" %(i))

V=np.zeros((len(Y)-50,50))
for i in range(len(Y)-50):
  V[i,:]=Y[i:i+50,3]
V=V.reshape((len(V),1,50))
Vnorm = np.divide(np.subtract(V, 53.5),28).reshape((len(V),1,50))

Arquivo 0 carregado
Arquivo 1 carregado
Arquivo 2 carregado
Arquivo 3 carregado
Arquivo 4 carregado
Arquivo 5 carregado
Arquivo 6 carregado
Arquivo 7 carregado
Arquivo 8 carregado
Arquivo 9 carregado
Arquivo 10 carregado
Arquivo 11 carregado
Arquivo 12 carregado
Arquivo 13 carregado
Arquivo 14 carregado
Arquivo 15 carregado
Arquivo 16 carregado
Arquivo 17 carregado
Arquivo 18 carregado
Arquivo 19 carregado
Arquivo 20 carregado
Arquivo 21 carregado
Arquivo 22 carregado
Arquivo 23 carregado
Arquivo 24 carregado
Arquivo 25 carregado
Arquivo 26 carregado
Arquivo 27 carregado
Arquivo 28 carregado
Arquivo 29 carregado
Arquivo 30 carregado
Arquivo 31 carregado
Arquivo 32 carregado
Arquivo 33 carregado
Arquivo 34 carregado


In [ ]:
n_base=129565-50
x = np.zeros((n_base,150,240,3),dtype='uint8')
y=np.zeros((n_base,4))
v=np.zeros((n_base,1,50))
vn=np.zeros((n_base,1,50))

soma_train=0
n=0
for i in range(0,35):
  #Carrega
  xi = np.load('/content/drive/My Drive/Novos_Dados/XTreino'+str(i)+'.npy')
  yi = np.load('/content/drive/My Drive/Novos_Dados/YTreino'+str(i)+'.npy')
  print("Arquivo ",i)
  if i==0:
    xi=xi[50:len(xi)]
    yi=yi[50:len(yi)]
  n=len(yi) 

  #Treino
  x[soma_train:soma_train + n]= xi
  y[soma_train:soma_train + n]= yi
  v[soma_train:soma_train + n]= V[soma_train:soma_train + n]
  vn[soma_train:soma_train + n]= Vnorm[soma_train:soma_train + n]

  #Att
  soma_train = soma_train + n

Arquivo  0
Arquivo  1
Arquivo  2
Arquivo  3
Arquivo  4
Arquivo  5
Arquivo  6
Arquivo  7
Arquivo  8
Arquivo  9
Arquivo  10
Arquivo  11
Arquivo  12
Arquivo  13
Arquivo  14
Arquivo  15
Arquivo  16
Arquivo  17
Arquivo  18
Arquivo  19
Arquivo  20
Arquivo  21
Arquivo  22
Arquivo  23
Arquivo  24
Arquivo  25
Arquivo  26
Arquivo  27
Arquivo  28
Arquivo  29
Arquivo  30
Arquivo  31
Arquivo  32
Arquivo  33
Arquivo  34


In [ ]:
X=np.load('/content/drive/My Drive/Novos_Dados/XTreino0.npy')
Y=np.load('/content/drive/My Drive/Novos_Dados/YTreino0.npy')
print("Arquivo 0 carregado")
for i in range(1,23):
    X=np.concatenate((X, np.load('/content/drive/My Drive/Novos_Dados/XTreino'+str(i)+'.npy')))
    Y=np.concatenate((Y, np.load('/content/drive/My Drive/Novos_Dados/YTreino'+str(i)+'.npy')))
    print("Arquivo %d carregado" %(i))

Arquivo 0 carregado
Arquivo 1 carregado
Arquivo 2 carregado
Arquivo 3 carregado
Arquivo 4 carregado
Arquivo 5 carregado
Arquivo 6 carregado
Arquivo 7 carregado
Arquivo 8 carregado
Arquivo 9 carregado
Arquivo 10 carregado
Arquivo 11 carregado
Arquivo 12 carregado
Arquivo 13 carregado
Arquivo 14 carregado
Arquivo 15 carregado
Arquivo 16 carregado
Arquivo 17 carregado
Arquivo 18 carregado
Arquivo 19 carregado
Arquivo 20 carregado
Arquivo 21 carregado
Arquivo 22 carregado


In [ ]:
X.shape

(82536, 150, 240, 3)

In [ ]:
Y.shape

(82536, 4)

## **Building VGG-16 model**

Imports Keras modules

In [ ]:
from keras import models, Input
from keras.layers import Input, Flatten, Dense, InputLayer, Dropout, Lambda
from keras.models import load_model, Model
from keras.layers import concatenate
from keras.applications.vgg16 import VGG16 

Loads pretrained VGG-16 model from Keras. This model was trained with ImageNet 

In [ ]:
VGG = VGG16(include_top = False,input_shape = (150,240,3),  weights='imagenet')





58892288/58889256 [==============================] - 2s 0us/step








In [ ]:
VGG.summary()

Model: "vgg16"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 150, 240, 3)       0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 150, 240, 64)      1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 150, 240, 64)      36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 75, 120, 64)       0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 75, 120, 128)      73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 75, 120, 128)      147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 37, 60, 128)       0     

Adds an input layer with 150x240x3 shape and a normalization layer before pretrained model and a flatten layer after it.

In [ ]:
input_shape = (150, 240, 3)

# Input
input_tensor = Input(shape=input_shape)
lamb = Lambda(lambda x: x/255., input_shape=input_shape, name='lambda_norm')(input_tensor)

# Xcpetion
x = VGG(lamb)
x = Flatten()(x)

#Model
conv = Model(input_tensor, x)

In [ ]:
conv.summary()

Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 150, 240, 3)       0         
_________________________________________________________________
lambda_norm (Lambda)         (None, 150, 240, 3)       0         
_________________________________________________________________
vgg16 (Model)                (None, 4, 7, 512)         14714688  
_________________________________________________________________
flatten_1 (Flatten)          (None, 14336)             0         
Total params: 14,714,688
Trainable params: 14,714,688
Non-trainable params: 0
_________________________________________________________________


## **Process data**

Process all images in built model and generate outputs of shape (14336,)

In [ ]:
x_pred = conv.predict(x)

In [ ]:
x_pred.shape

(129515, 14336)

Saves all processed data in numpy array to Google Drive folder

In [ ]:
np.save("/content/drive/My Drive/Novos_Dados/vgg_pred_avg/vgg_pred_avg.npy", x_pred)

In [ ]:
np.save('/content/drive/My Drive/Novos_Dados/vgg_pred_avg/y_vgg_pred_avg.npy',y)

In [ ]:
np.save('/content/drive/My Drive/Novos_Dados/vgg_pred_avg/vn_vgg_pred_avg.npy',vn)